posts/models.py

```diff
@@ -584,18 +590,52 @@ def updated_related_conditionals(self):
             ]
             for conditional in related_conditionals:
                 conditional.post.update_pseudo_materialized_fields()
-                print("Updated conditional in post: ", conditional.post)
 
     def update_pseudo_materialized_fields(self):
         self.set_scheduled_close_time()
         self.set_actual_close_time()
         self.set_scheduled_resolve_time()
         self.set_open_time()
         self.set_resolved()
+        self.update_global_leaderboard_tags()
         self.save()
         # Note: No risk of infinite loops since conditionals can't father other conditionals
         self.updated_related_conditionals()
 
+    def update_global_leaderboard_tags(self):
+        # set or update the tags for this post with respect to the global
+        # leaderboard(s) this is a part of
+
+        # Skip if post is not eligible for global leaderboards
+        if (
+            not self.default_project.type == Project.ProjectTypes.SITE_MAIN
+            and not self.projects.filter(type=Project.ProjectTypes.SITE_MAIN).exists()
+        ):
+            return
+
+        # Get all global leaderboard dates and create/get corresponding tags
+        to_set_tags = []
+        gl_dates = global_leaderboard_dates()
+        for question in self.get_questions():
+            dates = question.get_global_leaderboard_dates(gl_dates=gl_dates)
+            if dates:
+                tag_name, tag_slug = name_and_slug_for_global_leaderboard_dates(dates)
+                tag, _ = Project.objects.get_or_create(
+                    type=Project.ProjectTypes.TAG, name=tag_name, slug=tag_slug, order=1
+                )
+                to_set_tags.append(tag)
+
+        # Update post's global leaderboard tags
+        current_gl_tags = self.projects.filter_tags().filter(
```

Comment:
A small nit: `self.projects.filter` performs extra query. Post would almost never reach 5+ projects, so we can refactor these 2 filters to be performed on Python, not db level:
```
...
and not self.projects.filter(type=Project.ProjectTypes.SITE_MAIN).exists()
...
self.projects.filter_tags().filter(...)
```

by:

```
projects = post.projects.all()
...
# Check if has main
if next((p for p in projects if p.type ==Project.ProjectTypes.SITE_MAIN), None): ...

# Has tag
if next((p for p in projects if p.name.endswith == GLOBAL_LEADERBOARD_STRING), None): ...

```